client/lib/logmein/index.ts

```diff
@@ -0,0 +1,116 @@
+ * mapped domain enabling the nav bar and other features.
+ */
+/**
+ * External Dependencies
+ */
+import { Store } from 'redux';
+
+/**
+ * Internal Dependencies
+ */
+import getSitesItems from 'calypso/state/selectors/get-sites-items';
+import { isUserLoggedIn } from 'calypso/state/current-user/selectors';
+import { isEnabled } from '@automattic/calypso-config';
+
+// Used as placeholder / default domain to detect when we're looking at a relative url,
+// Note: also prevents exceptions from being raised
+const INVALID_URL = `https://__domain__.invalid`;
+
+type Host = string;
+
+let reduxStore: Store;
+
+export function logmeinUrl( url: string ): string {
+	// Disable feature if not enabled
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	// reduxStore must configured
+	if ( ! reduxStore ) {
+		return url;
+	}
+
+	// logmein=direct only works for logged into wordpress.com users
+	if ( ! isUserLoggedIn( reduxStore.getState() ) ) {
+		return url;
+	}
+
+	const newurl = new URL( String( url ), INVALID_URL );
+
+	// Ignore and passthrough invalid or /relative/urls that have no host or protocol specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	// Replace unmapped urls with their mapped url counterparts (*.wordpress.com => mapped.example.com)
+	const sites = Object.values( getSitesItems( reduxStore.getState() ) );
+
+	// We only want to logmein into valid sites that belong to the user (for now that is mapped simple sites)
+	// using INVALID_URL here to prevent the possibility of exceptions, if site.URL ever contains an invalid url
+	// the filtering will fail
+	const allowedHosts = sites
+		.map( ( site: any ) =>
+			isValidLogmeinSite( site ) ? new URL( site.URL, INVALID_URL ).host : false
+		)
+		.filter( Boolean );
+	if ( allowedHosts.indexOf( newurl.host ) === -1 ) {
+		return url;
+	}
+
+	// logmein doesn't work with http.
+	newurl.protocol = 'https:';
+
+	// Set the param
+	newurl.searchParams.set( 'logmein', 'direct' );
+
+	return newurl.toString();
+}
+
+function isValidLogmeinSite( site: any ): boolean {
+	return (
+		! site.is_vip &&
+		! site.jetpack &&
+		! site.options.is_automated_transfer &&
+		! site.options.is_domain_only &&
+		! site.options.is_redirect &&
+		! site.options.is_wpcom_store &&
+		site.options.is_mapped_domain
+	);
+}
+
+export function setLogmeinReduxStore( store: Store ): void {
+	reduxStore = store;
+}
+
+export function attachLogmein( store: Store ): void {
+	setLogmeinReduxStore( store );
+	document.addEventListener( 'click', logmeinOnClick, false );
+	document.addEventListener( 'contextmenu', logmeinOnRightClick, false );
+}
+
+const seen: Record< Host, boolean > = {};
+export function logmeinOnClick( event: MouseEvent ): void {
+	const link = ( event.target as HTMLElement ).closest( 'a' );
+
+	if ( link && link.href ) {
+		// Only apply logmein to each host once
+		const host = new URL( String( link.href ), INVALID_URL ).host;
+		if ( ! seen[ host ] ) {
```

Comment:
> This breaks if the link to the same host is re-rendered. Steps to reproduce:

This is the intention of the seen variable, the idea is to prevent people going through the logmein redirect flow more often than they need to. Once they're logged into the host they won't need to be logged in again for a while. Clearing cookies in between tabs is unlikely enough for it not to be an issue.

You can probably make the argument this is creating a flaky user experience and I'd agree but I think the trade off is worth it to prevent some additional redirects.

> The seen object is preserved between page redirects too, so there are other ways to reproduce this too. I missed the previous discussions here, so I'm curious to know its purpose 

Can you elaborate on this? My understanding is it being an in memory variable, shouldn't be cached between reloads - but pushstate navigation probably will survive.

> so I'm curious to know its purpose 

There was an earlier conversation in this PR with Niranjan about it but I can't track it down (rebases). There is no way to know if a user is logged in to a domain so we take this approach to reduce redirects, there will still be redirects that aren't needed but this is easier than tracking logged in state everywhere.

Reply:
> There was an earlier conversation in this PR with Niranjan about it but I can't track it down 
It was on [this thread](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r647586195).

Reply:
> This is the intention of the seen variable, the idea is to prevent people going through the logmein redirect flow more often than they need to. Once they're logged into the host they won't need to be logged in again for a while. Clearing cookies in between tabs is unlikely enough for it not to be an issue.

Makes sense! It slipped my mind that the link is modified on click and not on render, so we are good here!

> There is no way to know if a user is logged in to a domain so we take this approach to reduce redirects, there will still be redirects that aren't needed but this is easier than tracking logged in state everywhere.

Thanks! This does clear it up.